In [1]:
#Import Libraries
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import association_rules,apriori
import warnings
warnings.filterwarnings('ignore')

In [2]:
df = pd.read_excel("D:\DATA SCIENCE\ASSIGNMENT QUESTIONS\Association Rules\Online retail.xlsx",header=None)

In [3]:
df.columns = ['Products']

In [4]:
df

Products
0     shrimp,almonds,avocado,vegetables mix,green gr...
1                                burgers,meatballs,eggs
2                                               chutney
3                                        turkey,avocado
4     mineral water,milk,energy bar,whole wheat rice...
...                                                 ...
7496                      butter,light mayo,fresh bread
7497  burgers,frozen vegetables,eggs,french fries,ma...
7498                                            chicken
7499                                 escalope,green tea
7500    eggs,frozen smoothie,yogurt cake,low fat yogurt

[7501 rows x 1 columns]

In [9]:
null = df.isnull().any()
null

Products    False
dtype: bool

In [10]:
df.dtypes

Products    object
dtype: object

In [11]:
df.iloc[[0]]

Products
0  shrimp,almonds,avocado,vegetables mix,green gr...

In [12]:
def txt_split(txt):
    return txt.split(',')

In [13]:
txt_split("shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil")

['shrimp',
 'almonds',
 'avocado',
 'vegetables mix',
 'green grapes',
 'whole weat flour',
 'yams',
 'cottage cheese',
 'energy drink',
 'tomato juice',
 'low fat yogurt',
 'green tea',
 'honey',
 'salad',
 'mineral water',
 'salmon',
 'antioxydant juice',
 'frozen smoothie',
 'spinach',
 'olive oil']

In [14]:
data = df['Products'].apply(txt_split)
data

0       [shrimp, almonds, avocado, vegetables mix, gre...
1                              [burgers, meatballs, eggs]
2                                               [chutney]
3                                       [turkey, avocado]
4       [mineral water, milk, energy bar, whole wheat ...
                              ...                        
7496                    [butter, light mayo, fresh bread]
7497    [burgers, frozen vegetables, eggs, french frie...
7498                                            [chicken]
7499                                [escalope, green tea]
7500    [eggs, frozen smoothie, yogurt cake, low fat y...
Name: Products, Length: 7501, dtype: object

In [15]:
from mlxtend.preprocessing import TransactionEncoder

In [16]:
te = TransactionEncoder()
encoded_df = te.fit_transform(data)

In [17]:
encoded_df

array([[False,  True,  True, ...,  True, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       ...,
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False,  True, False]])

In [18]:
rules = pd.DataFrame(encoded_df,columns=te.columns_)

In [19]:
te.columns_

[' asparagus',
 'almonds',
 'antioxydant juice',
 'asparagus',
 'avocado',
 'babies food',
 'bacon',
 'barbecue sauce',
 'black tea',
 'blueberries',
 'body spray',
 'bramble',
 'brownies',
 'bug spray',
 'burger sauce',
 'burgers',
 'butter',
 'cake',
 'candy bars',
 'carrots',
 'cauliflower',
 'cereals',
 'champagne',
 'chicken',
 'chili',
 'chocolate',
 'chocolate bread',
 'chutney',
 'cider',
 'clothes accessories',
 'cookies',
 'cooking oil',
 'corn',
 'cottage cheese',
 'cream',
 'dessert wine',
 'eggplant',
 'eggs',
 'energy bar',
 'energy drink',
 'escalope',
 'extra dark chocolate',
 'flax seed',
 'french fries',
 'french wine',
 'fresh bread',
 'fresh tuna',
 'fromage blanc',
 'frozen smoothie',
 'frozen vegetables',
 'gluten free bar',
 'grated cheese',
 'green beans',
 'green grapes',
 'green tea',
 'ground beef',
 'gums',
 'ham',
 'hand protein bar',
 'herb & pepper',
 'honey',
 'hot dogs',
 'ketchup',
 'light cream',
 'light mayo',
 'low fat yogurt',
 'magazines',
 'mashe

In [20]:
rules

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0          False     True               True      False     True        False   
1          False    False              False      False    False        False   
2          False    False              False      False    False        False   
3          False    False              False      False     True        False   
4          False    False              False      False    False        False   
...          ...      ...                ...        ...      ...          ...   
7496       False    False              False      False    False        False   
7497       False    False              False      False    False        False   
7498       False    False              False      False    False        False   
7499       False    False              False      False    False        False   
7500       False    False              False      False    False        False   

      bacon  barbecue sauce  black tea  blueberries  ...  turkey  \
0     False           False      False        False  ...   False   
1     False           False      False        False  ...   False   
2     False           False      False        False  ...   False   
3     False           False      False        False  ...    True   
4     False           False      False        False  ...   False   
...     ...             ...        ...          ...  ...     ...   
7496  False           False      False        False  ...   False   
7497  False           False      False        False  ...   False   
7498  False           False      False        False  ...   False   
7499  False           False      False        False  ...   False   
7500  False           False      False        False  ...   False   

      vegetables mix  water spray  white wine  whole weat flour  \
0               True        False       False              True   
1              False        False       False             False   
2              False        False       False             False   
3              False        False       False             False   
4              False        False       False             False   
...              ...          ...         ...               ...   
7496           False        False       False             False   
7497           False        False       False             False   
7498           False        False       False             False   
7499           False        False       False             False   
7500           False        False       False             False   

      whole wheat pasta  whole wheat rice   yams  yogurt cake  zucchini  
0                 False             False   True        False     False  
1                 False             False  False        False     False  
2                 False             False  False        False     False  
3                 False             False  False        False     False  
4                 False              True  False        False     False  
...                 ...               ...    ...          ...       ...  
7496              False             False  False        False     False  
7497              False             False  False        False     False  
7498              False             False  False        False     False  
7499              False             False  False        False     False  
7500              False             False  False         True     False  

[7501 rows x 120 columns]

In [21]:
rules.replace([False,True],[0,1],inplace=True)

In [23]:
rules

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0              0        1                  1          0        1            0   
1              0        0                  0          0        0            0   
2              0        0                  0          0        0            0   
3              0        0                  0          0        1            0   
4              0        0                  0          0        0            0   
...          ...      ...                ...        ...      ...          ...   
7496           0        0                  0          0        0            0   
7497           0        0                  0          0        0            0   
7498           0        0                  0          0        0            0   
7499           0        0                  0          0        0            0   
7500           0        0                  0          0        0            0   

      bacon  barbecue sauce  black tea  blueberries  body spray  bramble  \
0         0               0          0            0           0        0   
1         0               0          0            0           0        0   
2         0               0          0            0           0        0   
3         0               0          0            0           0        0   
4         0               0          0            0           0        0   
...     ...             ...        ...          ...         ...      ...   
7496      0               0          0            0           0        0   
7497      0               0          0            0           0        0   
7498      0               0          0            0           0        0   
7499      0               0          0            0           0        0   
7500      0               0          0            0           0        0   

      brownies  bug spray  burger sauce  burgers  butter  cake  candy bars  \
0            0          0             0        0       0     0           0   
1            0          0             0        1       0     0           0   
2            0          0             0        0       0     0           0   
3            0          0             0        0       0     0           0   
4            0          0             0        0       0     0           0   
...        ...        ...           ...      ...     ...   ...         ...   
7496         0          0             0        0       1     0           0   
7497         0          0             0        1       0     0           0   
7498         0          0             0        0       0     0           0   
7499         0          0             0        0       0     0           0   
7500         0          0             0        0       0     0           0   

      carrots  cauliflower  cereals  champagne  chicken  chili  chocolate  \
0           0            0        0          0        0      0          0   
1           0            0        0          0        0      0          0   
2           0            0        0          0        0      0          0   
3           0            0        0          0        0      0          0   
4           0            0        0          0        0      0          0   
...       ...          ...      ...        ...      ...    ...        ...   
7496        0            0        0          0        0      0          0   
7497        0            0        0          0        0      0          0   
7498        0            0        0          0        1      0          0   
7499        0            0        0          0        0      0          0   
7500        0            0        0          0        0      0          0   

      chocolate bread  chutney  cider  clothes accessories  cookies  \
0                   0        0      0                    0        0   
1                   0        0      0                    0        0   
2                   0        1      0         

In [22]:
pd.set_option('display.max_columns', None)

In [24]:
scores = apriori(rules,min_support=0.01, use_colnames=True)

In [25]:
scores

support                                 itemsets
0    0.020397                                (almonds)
1    0.033329                                (avocado)
2    0.010799                         (barbecue sauce)
3    0.014265                              (black tea)
4    0.011465                             (body spray)
..        ...                                      ...
252  0.011065       (milk, ground beef, mineral water)
253  0.017064  (ground beef, mineral water, spaghetti)
254  0.015731         (milk, mineral water, spaghetti)
255  0.010265    (mineral water, spaghetti, olive oil)
256  0.011465     (mineral water, spaghetti, pancakes)

[257 rows x 2 columns]

In [26]:
score_rules = association_rules(scores,min_threshold=0.2)

In [27]:
score_rules

antecedents      consequents  antecedent support  \
0                     (avocado)  (mineral water)            0.033329   
1                     (burgers)           (eggs)            0.087188   
2                     (burgers)   (french fries)            0.087188   
3                     (burgers)      (green tea)            0.087188   
4                     (burgers)           (milk)            0.087188   
..                          ...              ...                 ...   
157  (mineral water, spaghetti)           (milk)            0.059725   
158  (mineral water, olive oil)      (spaghetti)            0.027596   
159      (spaghetti, olive oil)  (mineral water)            0.022930   
160   (mineral water, pancakes)      (spaghetti)            0.033729   
161       (spaghetti, pancakes)  (mineral water)            0.025197   

     consequent support   support  confidence      lift  leverage  conviction  \
0              0.238368  0.011598    0.348000  1.459926  0.003654    1.168147   
1              0.179709  0.028796    0.330275  1.837830  0.013128    1.224818   
2              0.170911  0.021997    0.252294  1.476173  0.007096    1.108844   
3              0.132116  0.017464    0.200306  1.516139  0.005945    1.085270   
4              0.129583  0.017864    0.204893  1.581175  0.006566    1.094717   
..                  ...       ...         ...       ...       ...         ...   
157            0.129583  0.015731    0.263393  2.032623  0.007992    1.181657   
158            0.174110  0.010265    0.371981  2.136468  0.005460    1.315071   
159            0.238368  0.010265    0.447674  1.878079  0.004799    1.378954   
160            0.174110  0.011465    0.339921  1.952333  0.005593    1.251198   
161            0.238368  0.011465    0.455026  1.908923  0.005459    1.397557   

     zhangs_metric  
0         0.325896  
1         0.499424  
2         0.353384  
3         0.372947  
4         0.402667  
..             ...  
157       0.540294  
158       0.547034  
159       0.478514  
160       0.504819  
161       0.488452  

[162 rows x 10 columns]

In [28]:
score_rules.sort_values(by = 'confidence', ascending=False)

antecedents          consequents  antecedent support  \
134         (ground beef, eggs)      (mineral water)            0.019997   
149         (milk, ground beef)      (mineral water)            0.021997   
121    (chocolate, ground beef)      (mineral water)            0.023064   
144   (milk, frozen vegetables)      (mineral water)            0.023597   
100                      (soup)      (mineral water)            0.050527   
..                          ...                  ...                 ...   
22               (french fries)          (chocolate)            0.170911   
146  (mineral water, spaghetti)  (frozen vegetables)            0.059725   
74                  (green tea)          (spaghetti)            0.132116   
34                       (soup)          (chocolate)            0.050527   
3                     (burgers)          (green tea)            0.087188   

     consequent support   support  confidence      lift  leverage  conviction  \
134            0.238368  0.010132    0.506667  2.125563  0.005365    1.543848   
149            0.238368  0.011065    0.503030  2.110308  0.005822    1.532552   
121            0.238368  0.010932    0.473988  1.988472  0.005434    1.447937   
144            0.238368  0.011065    0.468927  1.967236  0.005440    1.434136   
100            0.238368  0.023064    0.456464  1.914955  0.011020    1.401255   
..                  ...       ...         ...       ...       ...         ...   
22             0.163845  0.034395    0.201248  1.228284  0.006393    1.046827   
146            0.095321  0.011998    0.200893  2.107549  0.006305    1.132113   
74             0.174110  0.026530    0.200807  1.153335  0.003527    1.033405   
34             0.163845  0.010132    0.200528  1.223888  0.001853    1.045884   
3              0.132116  0.017464    0.200306  1.516139  0.005945    1.085270   

     zhangs_metric  
134       0.540342  
149       0.537969  
121       0.508837  
144       0.503555  
100       0.503221  
..             ...  
22        0.224169  
146       0.558895  
74        0.153188  
34        0.192667  
3         0.372947  

[162 rows x 10 columns]

We can see that if ground beef and eggs is purchased than mineral water will be purchased which is having 50% chance and also lift ratio 2.12 indicating both of them occurs
Similary if milk and ground beef is purchased then mineral water is also purchased with 50% confidence and lift 2.11

In [29]:
score_rules.sort_values(by = 'lift', ascending=False)

antecedents          consequents  antecedent support  \
75              (herb & pepper)        (ground beef)            0.049460   
154  (mineral water, spaghetti)        (ground beef)            0.059725   
69                   (tomatoes)  (frozen vegetables)            0.068391   
67                     (shrimp)  (frozen vegetables)            0.071457   
143       (milk, mineral water)  (frozen vegetables)            0.047994   
..                          ...                  ...                 ...   
74                  (green tea)          (spaghetti)            0.132116   
46                (ground beef)               (eggs)            0.098254   
19                  (chocolate)               (eggs)            0.163845   
73                  (green tea)      (mineral water)            0.132116   
57                   (escalope)      (mineral water)            0.079323   

     consequent support   support  confidence      lift  leverage  conviction  \
75             0.098254  0.015998    0.323450  3.291994  0.011138    1.332860   
154            0.098254  0.017064    0.285714  2.907928  0.011196    1.262445   
69             0.095321  0.016131    0.235867  2.474464  0.009612    1.183930   
67             0.095321  0.016664    0.233209  2.446574  0.009853    1.179825   
143            0.095321  0.011065    0.230556  2.418737  0.006490    1.175757   
..                  ...       ...         ...       ...       ...         ...   
74             0.174110  0.026530    0.200807  1.153335  0.003527    1.033405   
46             0.179709  0.019997    0.203528  1.132539  0.002340    1.029905   
19             0.179709  0.033196    0.202604  1.127397  0.003751    1.028711   
73             0.238368  0.031063    0.235116  0.986357 -0.000430    0.995748   
57             0.238368  0.017064    0.215126  0.902495 -0.001844    0.970387   

     zhangs_metric  
75        0.732460  
154       0.697788  
69        0.639616  
67        0.636767  
143       0.616132  
..             ...  
74        0.153188  
46        0.129779  
19        0.135143  
73       -0.015688  
57       -0.105024  

[162 rows x 10 columns]

(herb & pepper)	and (ground beef), (mineral water, spaghetti) and (ground beef) This 2 combination with lift 3.23 and 2.90 indictes that more than 3 times this products are purchased together